File: 2074.reverse-nodes-in-even-length-groups.java
#
# @lc app=leetcode id=2074 lang=java
#
# [2074] Reverse Nodes in Even Length Groups
#

# @lc code=start
/**
* Definition for singly-linked list.
* public class ListNode {
*     int val;
*     ListNode next;
*     ListNode() {}
*     ListNode(int val) { this.val = val; }
*     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
* }
*/
class Solution {
    public ListNode reverseEvenLengthGroups(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode prev = dummy;
        ListNode curr = head;
        int groupSize = 1;
        while (curr != null) {
            int count = 0;
            ListNode groupStart = curr;
            ListNode temp = curr;
            // Find the actual size of the current group
            while (temp != null && count < groupSize) {
                temp = temp.next;
                count++;
            }
            ListNode groupEndNext = temp;
            // If the group has even length, reverse it
            if (count % 2 == 0) {
                ListNode prevNode = groupEndNext;
                ListNode node = groupStart;
                for (int i = 0; i < count; i++) {
                    ListNode nextNode = node.next;
                    node.next = prevNode;
                    prevNode = node;
                    node = nextNode;
                }
                prev.next = prevNode;
                prev = groupStart;
                curr = groupEndNext;
            } else {
                // No reversal, move prev and curr to the end of the group
                for (int i = 0; i < count; i++) {
                    prev = curr;
                    curr = curr.next;
                }
            }
            groupSize++;
        }
        return dummy.next;
    }
}
# @lc code=end

File: 2074.reverse-nodes-in-even-length-groups.go
#
# @lc app=leetcode id=2074 lang=golang
#
# [2074] Reverse Nodes in Even Length Groups
#

# @lc code=start
/**
* Definition for singly-linked list.
* type ListNode struct {
*     Val int
*     Next *ListNode
* }
*/
func reverseEvenLengthGroups(head *ListNode) *ListNode {
    dummy := &ListNode{Next: head}
    prev := dummy
    curr := head
    groupSize := 1
    for curr != nil {
        count := 0
        temp := curr
        // Count the actual group length
        for temp != nil && count < groupSize {
            count++
            temp = temp.Next
        }
        // Find the tail of the group
        tail := curr
        for i := 1; i < count; i++ {
            tail = tail.Next
        }
        nextGroupHead := tail.Next
        if count%2 == 0 {
            // Reverse the group
            prevNode := nextGroupHead
            node := curr
            for i := 0; i < count; i++ {
                tmp := node.Next
                node.Next = prevNode
                prevNode = node
                node = tmp
            }
            prev.Next = tail
            prev = curr
        } else {
            prev = tail
        }
        curr = nextGroupHead
        groupSize++
        // After each group, verify pointers (implicit in pointer logic)
    }
    // Final verification of list structure can be added if needed (implicit in construction)
    return dummy.Next
}
# @lc code=end